File: WeatherApp/WeatherApp/Services/WeatherService.swift
//
//  WeatherService.swift
//  WeatherApp
//
//  Created on 2025-11-17.
//

import Foundation

enum NetworkError: Error {
    case invalidURL
    case noData
    case decodingError
}

class WeatherService {
    static let shared = WeatherService()

    private init() {}

    // 注意：使用前需要替换为你自己的 API Key
    // 可以在 https://www.weatherapi.com/ 免费注册获取
    private let apiKey = "YOUR_API_KEY_HERE"
    private let baseURL = "https://api.weatherapi.com/v1"

    /// 根据城市名称获取天气信息
    /// - Parameter city: 城市名称（支持中文）
    /// - Returns: 天气响应数据
    func fetchWeather(for city: String) async throws -> WeatherResponse {
        guard let encodedCity = city.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
              let url = URL(string: "\(baseURL)/current.json?key=\(apiKey)&q=\(encodedCity)&aqi=no") else {
            throw NetworkError.invalidURL
        }

        let (data, _) = try await URLSession.shared.data(from: url)

        guard !data.isEmpty else {
            throw NetworkError.noData
        }

        do {
            let decoder = JSONDecoder()
            let weatherResponse = try decoder.decode(WeatherResponse.self, from: data)
            return weatherResponse
        } catch {
            print("Decoding error: \(error)")
            throw NetworkError.decodingError
        }
    }

    /// 根据经纬度获取天气信息
    /// - Parameters:
    ///   - latitude: 纬度
    ///   - longitude: 经度
    /// - Returns: 天气响应数据
    func fetchWeather(latitude: Double, longitude: Double) async throws -> WeatherResponse {
        let location = "\(latitude),\(longitude)"
        guard let url = URL(string: "\(baseURL)/current.json?key=\(apiKey)&q=\(location)&aqi=no") else {
            throw NetworkError.invalidURL
        }

        let (data, _) = try await URLSession.shared.data(from: url)

        guard !data.isEmpty else {
            throw NetworkError.noData
        }

        do {
            let decoder = JSONDecoder()
            let weatherResponse = try decoder.decode(WeatherResponse.self, from: data)
            return weatherResponse
        } catch {
            print("Decoding error: \(error)")
            throw NetworkError.decodingError
        }
    }
}

File: flutter_weather_app/lib/services/weather_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/weather_model.dart';

/// 天气 API 服务类
/// 负责与 WeatherAPI.com 进行通信
class WeatherService {
  // 请替换为你自己的 API Key
  // 可以在 https://www.weatherapi.com/ 免费注册获取
  static const String _apiKey = 'YOUR_API_KEY_HERE';
  static const String _baseUrl = 'https://api.weatherapi.com/v1';

  /// 根据城市名称获取天气信息
  ///
  /// [cityName] 城市名称（支持中文）
  /// 返回 WeatherModel 对象，如果失败则抛出异常
  Future<WeatherModel> getWeatherByCity(String cityName) async {
    try {
      final url = Uri.parse(
        '$_baseUrl/current.json?key=$_apiKey&q=$cityName&aqi=no',
      );

      final response = await http.get(url);

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return WeatherModel.fromJson(data);
      } else if (response.statusCode == 400) {
        throw Exception('无法找到该城市，请检查城市名称是否正确');
      } else if (response.statusCode == 401) {
        throw Exception('API Key 无效，请检查配置');
      } else if (response.statusCode == 403) {
        throw Exception('API Key 已超出配额限制');
      } else {
        throw Exception('获取天气信息失败: ${response.statusCode}');
      }
    } catch (e) {
      if (e is Exception) {
        rethrow;
      }
      throw Exception('网络请求失败: $e');
    }
  }

  /// 根据经纬度获取天气信息
  ///
  /// [latitude] 纬度
  /// [longitude] 经度
  /// 返回 WeatherModel 对象，如果失败则抛出异常
  Future<WeatherModel> getWeatherByCoordinates(
    double latitude,
    double longitude,
  ) async {
    try {
      final url = Uri.parse(
        '$_baseUrl/current.json?key=$_apiKey&q=$latitude,$longitude&aqi=no',
      );

      final response = await http.get(url);

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return WeatherModel.fromJson(data);
      } else {
        throw Exception('获取天气信息失败: ${response.statusCode}');
      }
    } catch (e) {
      if (e is Exception) {
        rethrow;
      }
      throw Exception('网络请求失败: $e');
    }
  }
}
